File: ai-bot/services/gateway/internal/ship/stacks_test.go
// Testes do REGISTRO.
//
// Não é teste de tautologia: são os valores que o pipeline usa de verdade, e
// um erro de transcrição aqui manda o build para o diretório errado sem
// reclamar. Foi por isso que a porta original testava os valores, e a porta da
// porta testa de novo.
package ship

import (
	"encoding/json"
	"regexp"
	"strings"
	"testing"
)

func mustStack(t *testing.T, id string) Stack {
	t.Helper()
	stack, ok := StackByID(id)
	if !ok {
		t.Fatalf("a stack %q não existe no registro", id)
	}
	return stack
}

func TestCatalogCarriesTheWholePort(t *testing.T) {
	if len(StackIDs) != 47 {
		t.Fatalf("o port trouxe 47 stacks; StackIDs tem %d", len(StackIDs))
	}
	if len(Stacks) != 47 {
		t.Fatalf("o port trouxe 47 stacks; o mapa tem %d", len(Stacks))
	}
	if len(Languages) != 11 {
		t.Fatalf("o port trouxe 11 linguagens; o mapa tem %d", len(Languages))
	}
}

// StackIDs é uma lista PARALELA ao mapa — e lista paralela deriva. Este teste
// impede a stack registrada que nunca é varrida e o id varrido que não existe.
func TestStackIDsMatchTheMapExactly(t *testing.T) {
	seen := make(map[string]bool, len(StackIDs))
	for _, id := range StackIDs {
		if seen[id] {
			t.Errorf("id duplicado em StackIDs: %q", id)
		}
		seen[id] = true
		if _, ok := Stacks[id]; !ok {
			t.Errorf("StackIDs lista %q, que não existe no mapa", id)
		}
	}
	for id := range Stacks {
		if !seen[id] {
			t.Errorf("a stack %q está no mapa e fora de StackIDs — a detecção nunca a varre", id)
		}
	}
}

func TestEveryStackPointsToAnExistingLanguage(t *testing.T) {
	for id, stack := range Stacks {
		if _, ok := LanguageByID(stack.Language); !ok {
			t.Errorf("stack %s aponta para a linguagem %q, que não existe", id, stack.Language)
		}
	}
}

func TestEveryStackHasPortAndOutputDirectory(t *testing.T) {
	for id, stack := range Stacks {
		if stack.DefaultPort <= 0 {
			t.Errorf("stack %s sem porta padrão", id)
		}
		if stack.OutputDirectory == "" {
			t.Errorf("stack %s sem diretório de saída", id)
		}
		if stack.Name == "" {
			t.Errorf("stack %s sem nome", id)
		}
	}
}

func TestEverydayTruthValues(t *testing.T) {
	// Os casos que mais aparecem no dia. Um dígito trocado aqui e o deploy do
	// Next procura o build em `.dist` para sempre.
	if got := mustStack(t, "nextjs").OutputDirectory; got != ".next" {
		t.Errorf("nextjs sai em %q, esperava .next", got)
	}
	if got := mustStack(t, "nextjs").DefaultBuildCommand; got != "next build" {
		t.Errorf("build do nextjs é %q, esperava next build", got)
	}
	if got := mustStack(t, "nuxt").DefaultStartCommand; got != "node .output/server/index.mjs" {
		t.Errorf("start do nuxt é %q, esperava node .output/server/index.mjs", got)
	}
	if got := mustStack(t, "vite").DefaultPort; got != 5173 {
		t.Errorf("porta do vite é %d, esperava 5173", got)
	}
	if got := mustStack(t, "astro").DefaultPort; got != 4321 {
		t.Errorf("porta do astro é %d, esperava 4321", got)
	}
	if got := mustStack(t, "django").DefaultStartCommand; !strings.Contains(got, "gunicorn") {
		t.Errorf("start do django é %q, esperava gunicorn", got)
	}
	if got := mustStack(t, "go").DefaultPort; got != 8080 {
		t.Errorf("porta do go é %d, esperava 8080", got)
	}
	if got := mustStack(t, "rust").OutputDirectory; got != "target/release" {
		t.Errorf("rust sai em %q, esperava target/release", got)
	}
}

func TestRailsDeclaresGemfileAndAConfirmer(t *testing.T) {
	// A regra irregular do original: só `Gemfile` casaria com qualquer
	// projeto Ruby, e o Sinatra viraria Rails. A conjunção em si é testada em
	// detect_test.go; aqui se garante que os DADOS dela existem.
	rails := mustStack(t, "rails")
	if rails.Detection == nil {
		t.Fatal("rails sem bloco de detecção")
	}
	markers := strings.Join(rails.Detection.RootMarkers, " ")
	if !strings.Contains(markers, "Gemfile") {
		t.Error("os marcadores do rails não trazem Gemfile")
	}
	if !strings.Contains(markers, "config/routes.rb") {
		t.Error("os marcadores do rails não trazem config/routes.rb")
	}
}

func TestCRAIdentifiesByDependencyNotLayout(t *testing.T) {
	// `public/` + `src/` é layout de meio mundo; `react-scripts` é o sinal.
	cra := mustStack(t, "cra")
	if cra.Detection == nil {
		t.Fatal("cra sem bloco de detecção")
	}
	found := false
	for _, dep := range cra.Detection.Deps {
		if dep == "react-scripts" {
			found = true
		}
	}
	if !found {
		t.Error("o cra não se identifica por react-scripts")
	}
	if len(cra.Detection.RootMarkers) != 0 {
		t.Errorf("o cra não deveria ter marcador de raiz, tem %v", cra.Detection.RootMarkers)
	}
}

func TestRootMarkerIndexCoversKnownAnchors(t *testing.T) {
	for _, marker := range []string{"next.config.js", "go.mod", "Cargo.toml", "manage.py", "Dockerfile", "artisan"} {
		if !StackRootMarkers[marker] {
			t.Errorf("o índice de marcadores não cobre %q", marker)
		}
	}
}

func TestNeverTransfersWhatInstallRebuilds(t *testing.T) {
	excludes := strings.Join(TransferExcludes, " ")
	if !strings.Contains(excludes, "node_modules") {
		t.Error("TransferExcludes não poda node_modules")
	}
	if !strings.Contains(excludes, ".git") {
		t.Error("TransferExcludes não poda .git")
	}
	// `build` e `dist` também são podados, mas SÓ na raiz do pacote.
	want := []string{"build", "dist", "data"}
	if len(PackageRootOnlyExcludes) != len(want) {
		t.Fatalf("PackageRootOnlyExcludes é %v, esperava %v", PackageRootOnlyExcludes, want)
	}
	for index, value := range want {
		if PackageRootOnlyExcludes[index] != value {
			t.Fatalf("PackageRootOnlyExcludes é %v, esperava %v", PackageRootOnlyExcludes, want)
		}
	}
}

func TestNoBrandLogosAndNoExternalURLs(t *testing.T) {
	// O original trazia STACK_ICONS apontando para um CDN. Buscar logo de
	// marca lá fora entrega a um terceiro a lista de frameworks que o usuário
	// tem, quebra o app offline e depende de uma URL móvel (`@latest`).
	raw, err := json.Marshal(Stacks)
	if err != nil {
		t.Fatalf("serializar o registro: %v", err)
	}
	text := string(raw)
	if strings.Contains(text, "jsdelivr") || strings.Contains(text, "devicon") {
		t.Fatal("o registro voltou a carregar ícone de CDN")
	}
	// `http://0.0.0.0` é bind local do .NET, não busca de rede: o que não
	// pode aparecer é host de terceiro.
	urls := regexp.MustCompile(`https?://[^"\\ ]+`).FindAllString(text, -1)
	local := regexp.MustCompile(`^https?://(0\.0\.0\.0|127\.0\.0\.1|localhost)`)
	for _, url := range urls {
		if !local.MatchString(url) {
			t.Errorf("URL externa dentro do registro: %s", url)
		}
	}
}

func TestImageFallsBackToTheLanguage(t *testing.T) {
	nextjs := mustStack(t, "nextjs")
	if nextjs.BuildImage != "" {
		t.Fatalf("o nextjs não deveria sobrescrever a imagem de build, tem %q", nextjs.BuildImage)
	}
	typescript, ok := LanguageByID("typescript")
	if !ok {
		t.Fatal("a linguagem typescript sumiu do registro")
	}
	if got := BuildImageFor(nextjs); got != typescript.BuildImage {
		t.Errorf("BuildImageFor(nextjs) = %q, esperava %q", got, typescript.BuildImage)
	}
	if got := RuntimeImageFor(nextjs); got != typescript.RuntimeImage {
		t.Errorf("RuntimeImageFor(nextjs) = %q, esperava %q", got, typescript.RuntimeImage)
	}
}

func TestDefaultInstallCommandPerLanguage(t *testing.T) {
	cases := map[string]string{
		"nextjs":  "npm install",
		"go":      "go mod download",
		"laravel": "composer install",
		// Instalação embutida no build: devolver vazio é correto.
		"rails":  "",
		"python": "",
		"dotnet": "",
	}
	for id, want := range cases {
		if got := DefaultInstallCommand(mustStack(t, id)); got != want {
			t.Errorf("DefaultInstallCommand(%s) = %q, esperava %q", id, got, want)
		}
	}
}

File: apps/desktop/src/lib/ship/stacks.test.ts
import { describe, expect, it } from "vitest";

import {
  buildImageFor,
  getLanguage,
  getStack,
  LANGUAGES,
  PACKAGE_ROOT_ONLY_EXCLUDES,
  runtimeImageFor,
  STACKS,
  STACK_IDS,
  STACK_ROOT_MARKERS,
  TRANSFER_EXCLUDES,
  type StackDefinition
} from "./stacks";
import { formatBuildEvent, generateDockerfile, parseBuildEvent, progressForStep } from "./dockerfile";

describe("registro de stacks", () => {
  it("trouxe o catálogo inteiro do port", () => {
    expect(STACK_IDS.length).toBeGreaterThanOrEqual(45);
    expect(Object.keys(LANGUAGES).length).toBeGreaterThanOrEqual(11);
  });

  it("toda stack aponta para uma linguagem que existe", () => {
    for (const [id, stack] of Object.entries(STACKS as Record<string, StackDefinition>)) {
      expect(getLanguage(stack.language), `stack ${id} → linguagem ${stack.language}`).toBeTruthy();
    }
  });

  it("toda stack tem porta e diretório de saída", () => {
    for (const [id, stack] of Object.entries(STACKS as Record<string, StackDefinition>)) {
      expect(typeof stack.outputDirectory, id).toBe("string");
      expect(stack.defaultPort, id).toBeGreaterThan(0);
    }
  });

  it("os casos que mais aparecem no dia estão certos", () => {
    // Não é teste de tautologia: são os valores que o pipeline usa de verdade,
    // e um erro de transcrição aqui manda o build para o diretório errado.
    expect(getStack("nextjs")?.outputDirectory).toBe(".next");
    expect(getStack("nextjs")?.defaultBuildCommand).toBe("next build");
    expect(getStack("nuxt")?.defaultStartCommand).toBe("node .output/server/index.mjs");
    expect(getStack("vite")?.defaultPort).toBe(5173);
    expect(getStack("astro")?.defaultPort).toBe(4321);
    expect(getStack("django")?.defaultStartCommand).toContain("gunicorn");
  });

  it("o Rails exige o Gemfile E um confirmador", () => {
    // A regra irregular do original: só `Gemfile` casaria com qualquer projeto
    // Ruby, e o Sinatra viraria Rails.
    const rails = getStack("rails");
    expect(rails?.detection?.rootMarkers).toContain("Gemfile");
    expect(rails?.detection?.rootMarkers).toContain("config/routes.rb");
  });

  it("o CRA se identifica pela dependência, não pelo layout", () => {
    // `public/` + `src/` é layout de meio mundo; `react-scripts` é o sinal.
    expect(getStack("cra")?.detection?.deps).toContain("react-scripts");
    expect(getStack("cra")?.detection?.rootMarkers).toBeUndefined();
  });

  it("o índice de marcadores cobre os arquivos-âncora conhecidos", () => {
    for (const marcador of ["next.config.js", "go.mod", "Cargo.toml", "manage.py", "Dockerfile", "artisan"]) {
      expect(STACK_ROOT_MARKERS.has(marcador), marcador).toBe(true);
    }
  });

  it("nunca transfere o que o install refaz do outro lado", () => {
    expect(TRANSFER_EXCLUDES).toContain("node_modules");
    expect(TRANSFER_EXCLUDES).toContain(".git");
    // `build` e `dist` também são podados, mas SÓ na raiz do pacote.
    expect(PACKAGE_ROOT_ONLY_EXCLUDES).toEqual(["build", "dist", "data"]);
  });

  it("nenhum logo de marca veio junto — e nenhuma URL para fora", () => {
    // O original trazia `STACK_ICONS` apontando para um CDN. Buscar logo de
    // marca lá fora entrega a um terceiro a lista de frameworks que o usuário
    // tem, quebra o app offline e depende de uma URL móvel (`@latest`).
    const bruto = JSON.stringify(STACKS);
    expect(bruto).not.toContain("jsdelivr");
    expect(bruto).not.toContain("devicon");
    // `http://0.0.0.0` é bind local do .NET, não busca de rede: o que não pode
    // aparecer é host de terceiro.
    const urls = bruto.match(/https?:\/\/[^"\\]+/g) ?? [];
    const externas = urls.filter((url) => !/^https?:\/\/(0\.0\.0\.0|127\.0\.0\.1|localhost)/.test(url));
    expect(externas).toEqual([]);
  });

  it("a imagem cai na da linguagem quando a stack não sobrescreve", () => {
    const nextjs = getStack("nextjs")!;
    expect(nextjs.buildImage).toBeUndefined();
    expect(buildImageFor(nextjs)).toBe(LANGUAGES.typescript.buildImage);
    expect(runtimeImageFor(nextjs)).toBe(LANGUAGES.typescript.runtimeImage);
  });
});

describe("Dockerfile gerado", () => {
  it("multi-estágio quando build e runtime diferem, e copia só o que produz", () => {
    const go = getStack("go")!;
    const saida = generateDockerfile({
      stack: go,
      installCommand: "go mod download",
      buildCommand: "go build -o app ."
    });
    expect(saida).toContain("AS builder");
    expect(saida).toContain("AS runtime");
    // `productionPaths` do Go é ["app"] — o binário, não a árvore inteira.
    expect(saida).toContain("COPY --from=builder /workspace/app /app/app");
    expect(saida).toContain("EXPOSE 8080");
  });

  it("estágio único quando as imagens são a mesma", () => {
    const saida = generateDockerfile({ stack: getStack("nextjs")!, buildCommand: "next build" });
    expect(saida).not.toContain("AS builder");
    expect(saida).toContain("EXPOSE 3000");
    expect(saida).toContain('CMD ["sh", "-c", "next start"]');
  });

  it("install e build ficam num RUN só, com as marcas de progresso", () => {
    const saida = generateDockerfile({
      stack: getStack("nextjs")!,
      installCommand: "npm ci",
      buildCommand: "next build"
    });
    const runs = saida.split("\n").filter((linha) => linha.startsWith("RUN "));
    expect(runs).toHaveLength(1);
    expect(runs[0]).toContain("step=install status=running");
    expect(runs[0]).toContain("step=install status=completed");
    expect(runs[0]).toContain("step=build status=completed");
  });

  it("NO_COLOR entra sozinho e TERM fica de fora", () => {
    const saida = generateDockerfile({
      stack: getStack("nextjs")!,
      buildCommand: "next build",
      env: { TERM: "xterm-256color", API_URL: "https://exemplo" }
    });
    expect(saida).toContain("export NO_COLOR='1'");
    expect(saida).toContain("export API_URL='https://exemplo'");
    expect(saida).not.toContain("TERM=");
  });

  it("aspas simples no valor não escapam do shell", () => {
    const saida = generateDockerfile({
      stack: getStack("nextjs")!,
      buildCommand: "next build",
      env: { NOTA: "o'brien" }
    });
    expect(saida).toContain(`export NOTA='o'\\''brien'`);
  });

  it("subpasta de monorepo vira o WORKDIR do build", () => {
    const saida = generateDockerfile({
      stack: getStack("nextjs")!,
      sourceDir: "apps/web",
      buildCommand: "next build"
    });
    expect(saida).toContain("WORKDIR /workspace/apps/web");
  });

  it("sem comando de start não inventa CMD", () => {
    const saida = generateDockerfile({ stack: getStack("vite")!, buildCommand: "vite build" });
    expect(saida).not.toContain("CMD");
  });
});

describe("marcas de progresso", () => {
  it("a marca vai e volta", () => {
    const linha = formatBuildEvent("build", "completed");
    expect(parseBuildEvent(linha)).toEqual({ step: "build", status: "completed" });
  });

  it("linha comum não vira marca", () => {
    expect(parseBuildEvent("npm warn deprecated")).toBeNull();
    expect(parseBuildEvent("")).toBeNull();
  });

  it("o passo concluído vale dez pontos a mais", () => {
    expect(progressForStep("install")).toBe(30);
    expect(progressForStep("install", "completed")).toBe(40);
    expect(progressForStep("passo-que-nao-existe")).toBe(0);
  });
});
